### PR TITLE
Theme support to declare image sizes

### LINKED ISSUE #17557: Theming: Improve how product image sizes are defined by themes
This is a combination of https://github.com/woocommerce/woocommerce/issues/11666 and https://github.com/woocommerce/woocommerce/issues/12878 which overlap somewhat.

See https://github.com/woocommerce/woocommerce/pull/17610 for explanation.

- [x] For default themes we support out of the box, we should register the settings ourselves.
- [x] These settings should also remain filterable.
- [x] If a theme does not declare wc support, keep the image size settings visible
- [x] We should also review sizes naming and ensure it still makes sense.
- [x] Update tests
- [x] Update API
- [x] Upgrade routine

### PR DESCRIPTION
**Pull request by mikejolley**
This closes #17557 

Allows image sizes to be defined using add_theme_support. If used, image settings in WooCommerce are hidden. Includes support for all default WP themes.

Themes can pass in the image **widths** they want to use. e.g.

```
add_theme_support( 'woocommerce', array(
     'thumbnail_image_width' => 150,
     'single_image_width' => 322,
) );
```

If present, these settings take priority over settings, and settings are hidden.

There is a new setting called 'thumbnail cropping' which then allows the end user to select:

- Square crop (1:1)
- Uncropped
- Custom aspect ratio

Sizes are then calculated from the given widths.

**Single product images** are uncropped and not affected by ratio settings.

**After changing settings, regenerate thumbnails needs to run** or the changes will not be visible. https://github.com/woocommerce/woocommerce/issues/12880 will handle this.

### CODE REVIEW

File: includes/admin/class-wc-admin-notices.php [hunk 1]
```diff
@@ -79,7 +79,7 @@ public static function remove_all_notices() {
 	 * Reset notices for themes when switched or a new version of WC is installed.
 	 */
 	public static function reset_admin_notices() {
-		if ( ! current_theme_supports( 'woocommerce' ) && ! in_array( get_option( 'template' ), wc_get_core_supported_themes() ) ) {
+		if ( ! current_theme_supports( 'woocommerce' ) ) {
 			self::add_notice( 'theme_support' );
 		}
```
Comment (claudiulodro) [inline includes/admin/class-wc-admin-notices.php]: I think we can probably remove this notice entirely if all themes should work with WooCommerce.

Comment (claudiulodro) [reply includes/admin/class-wc-admin-notices.php]: Doesn't need to be in this PR. Just observing.

**Comment by claudiulodro** [comment]
This looks good to me. I like the new Twenty_X classes, too. I'll give it a little while for other people to review it then merge if nobody brings up anything.

**Comment by jameskoster** [comment]
The preview looks a bit odd in larger and smaller screens. Let's just move it below the settings;

preview

Overall it works great. I'll tweak the preview styling once you've moved it.

**Pull request merged**